qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -271,116 +323,200 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
             (for instance, when the specified synthesis method is not available).
         """
         qubits = tuple(dag.find_bit(q).index for q in dag.qubits)
+        context = QubitContext({i: i for i in range(len(dag.qubits))})
+        tracker = QubitTracker(num_qubits=dag.num_qubits())
         if self.qubits_initially_zero:
-            clean, dirty = set(qubits), set()
-        else:
-            clean, dirty = set(), set(qubits)
+            tracker.set_clean(context.to_globals(qubits))
 
-        tracker = QubitTracker(qubits=qubits, clean=clean, dirty=dirty)
-        return self._run(dag, tracker)
+        out_dag = self._run(dag, tracker, context, use_ancillas=True, top_level=True)
+        return out_dag
 
-    def _run(self, dag: DAGCircuit, tracker: QubitTracker) -> DAGCircuit:
-        # Check if HighLevelSynthesis can be skipped.
-        for node in dag.op_nodes():
-            qubits = tuple(dag.find_bit(q).index for q in node.qargs)
-            if not self._definitely_skip_node(node, qubits, dag):
-                break
-        else:
-            # The for-loop terminates without reaching the break statement
-            return dag
+    def _run(
+        self,
+        dag: DAGCircuit,
+        tracker: QubitTracker,
+        context: QubitContext,
+        use_ancillas: bool,
+        top_level: bool,
+    ) -> DAGCircuit:
+        """
+        The main recursive function that synthesizes a DAGCircuit.
+
+        Input:
+            dag: the DAG to be synthesized.
+            tracker: the global tracker, tracking the state of original qubits.
+            context: the correspondence between the dag's qubits and the global qubits.
+            use_ancillas: if True, synthesis algorithms are allowed to use ancillas.
+            top_level: specifies if this is the top-level of the recursion.
+
+        The function returns the synthesized DAG.
 
-        # Start by analyzing the nodes in the DAG. This for-loop is a first version of a potentially
-        # more elaborate approach to find good operation/ancilla allocations. It greedily iterates
-        # over the nodes, checking whether we can synthesize them, while keeping track of the
-        # qubit states. It does not trade-off allocations and just gives all available qubits
-        # to the current operation (a "the-first-takes-all" approach).
+        Note that by using the auxiliary qubits to synthesize operations present in the input DAG,
+        the synthesized DAG may be defined over more qubits than the input DAG. In this case,
+        the function update in-place the global qubits tracker and extends the local-to-global
+        context.
+        """
+
+        if dag.num_qubits() != context.num_qubits():
+            raise TranspilerError("HighLevelSynthesis internal error.")
```

Comment:
If there are no bugs, this should never-ever happen. However, this was very useful during PR implementation / debugging (as it allows to pinpoint the existence of a problem as early as possible), and I think keeping these checks would be good in case there are some bugs / unexpected corner-cases.  I don't know if I can think of a more meaningful message than will help debugging, modulo the fact that something is going wrong.